plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java

```diff
@@ -20,61 +20,127 @@
+                .expireAfterWrite(caseInsensitiveNameMatchingCacheTtl.toMillis(), MILLISECONDS);
+        this.remoteDatasetNames = remoteNamesCacheBuilder.build();
+        this.remoteTableNames = remoteNamesCacheBuilder.build();
+    }
+
+    String toRemoteDatasetName(String projectId, String datasetName)
+    {
+        requireNonNull(projectId, "projectId is null");
+        requireNonNull(datasetName, "datasetName is null");
+        verify(datasetName.codePoints().noneMatch(Character::isUpperCase), "Expected schema name from internal metadata to be lowercase: %s", datasetName);
+        if (caseInsensitiveNameMatching) {
+            try {
+                @Nullable String remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
+                if (remoteDatasetName == null) {
+                    // this might have been just created, reload cache
+                    // will throw on collisions to avoid ambiguity
+                    Map<String, String> mapping = stream(listDatasets(projectId))
+                            // will throw on collisions to avoid ambiguity
+                            .collect(toImmutableMap(dataset -> dataset.getDatasetId().getDataset().toLowerCase(ENGLISH), dataset -> dataset.getDatasetId().getDataset()));
+                    remoteDatasetNames.putAll(mapping);
+                    remoteDatasetName = mapping.get(datasetName);
+                }
+                if (remoteDatasetName != null) {
+                    return remoteDatasetName;
+                }
+            }
+            catch (RuntimeException e) {
+                throw new TrinoException(BIGQUERY_OBJECT_NOT_FOUND, "Failed to find remote dataset name: " + firstNonNull(e.getMessage(), e), e);
+            }
+        }
+
+        return datasetName;
     }
 
-    TableInfo getTable(TableId tableId)
+    String toRemoteTableName(String projectId, String remoteDatasetName, String tableName)
     {
-        TableId bigQueryTableId = tableIds.get(tableId);
-        Table table = bigQuery.getTable(bigQueryTableId != null ? bigQueryTableId : tableId);
-        if (table != null) {
-            tableIds.putIfAbsent(tableId, table.getTableId());
-            datasetIds.putIfAbsent(toDatasetId(tableId), toDatasetId(table.getTableId()));
+        requireNonNull(projectId, "projectId is null");
+        requireNonNull(remoteDatasetName, "remoteDatasetName is null");
+        requireNonNull(tableName, "tableName is null");
+        verify(tableName.codePoints().noneMatch(Character::isUpperCase), "Expected table name from internal metadata to be lowercase: %s", tableName);
+        if (caseInsensitiveNameMatching) {
+            try {
+                @Nullable Map<String, String> mapping = remoteTableNames.getIfPresent(remoteDatasetName);
+                if (mapping != null && !mapping.containsKey(tableName)) {
+                    // this might have been just created, reload cache
+                    mapping = null;
```

Comment:
I am a bit concerned about the fact that we are listing all the tables on each request for an inexistent table.
I think better approach would be to store individual tables in cache, each using separate key.
And also cache negative information. 

A cache storying lower_case_table_name_with_database -> Optional<remote_table_name> would work here.

Reply:
Negative caching comment also applies to `remoteDatasetNames` cache

Reply:
Good point about caching negative results. In current case it causes a cache miss everytime.

Changing the cache to a `Cache<TableId, TableId>` should resolve this point and can also allow us to retrieve only the tables in the provided dataset instead of listing all tables. WDYT?